File: Apps/MobileHarmonica/Domain/Entities/RichterTuning.swift
import Foundation

struct RichterTuning {
    // MARK: - Public

    func tone(for reed: Reed, in key: HarmonicaKey) -> Tone {
        Tone(
            pitch: note(for: reed).transposed(by: key.semitonesFromC).pitch,
            bendableSemitones: bendableSemitones(for: reed)
        )
    }

    // MARK: - Private

    private func bendableSemitones(for reed: Reed) -> Double {
        let bending = note(for: reed).number
        let neighbour = note(for: Reed(hole: reed.hole, breath: reed.breath.reversed)).number
        guard bending > neighbour else { return 0 }

        return Double(bending - neighbour - 1)
    }

    private func note(for reed: Reed) -> MIDINote {
        switch reed.breath {
        case .blow: blowNote(for: reed.hole)
        case .draw: drawNote(for: reed.hole)
        }
    }

    private func blowNote(for hole: Hole) -> MIDINote {
        switch hole {
        case .one: .c4
        case .two: .e4
        case .three: .g4
        case .four: .c5
        case .five: .e5
        case .six: .g5
        case .seven: .c6
        case .eight: .e6
        case .nine: .g6
        case .ten: .c7
        }
    }

    private func drawNote(for hole: Hole) -> MIDINote {
        switch hole {
        case .one: .d4
        case .two: .g4
        case .three: .b4
        case .four: .d5
        case .five: .f5
        case .six: .a5
        case .seven: .b5
        case .eight: .d6
        case .nine: .f6
        case .ten: .a6
        }
    }
}

File: Web/src/domain/instrument/richterTuning.ts
import { reversed } from './breath.js'
import type { Hole } from './hole.js'
import type { HarmonicaKey } from './harmonicaKey.js'
import { semitonesFromC } from './harmonicaKey.js'
import type { MidiNote } from './midiNote.js'
import { transposedBy } from './midiNote.js'
import type { Reed } from './reed.js'

const blowNotes: Record<Hole, MidiNote> = {
    1: 60, 2: 64, 3: 67, 4: 72, 5: 76, 6: 79, 7: 84, 8: 88, 9: 91, 10: 96
}

const drawNotes: Record<Hole, MidiNote> = {
    1: 62, 2: 67, 3: 71, 4: 74, 5: 77, 6: 81, 7: 83, 8: 86, 9: 89, 10: 93
}

export function noteFor(reed: Reed, key: HarmonicaKey): MidiNote {
    return transposedBy(plainNote(reed), semitonesFromC(key))
}

export function bendableSemitones(reed: Reed): number {
    const bending = plainNote(reed)
    const neighbour = plainNote({ hole: reed.hole, breath: reversed(reed.breath) })
    if (bending <= neighbour) return 0

    return bending - neighbour - 1
}

export function overbendableSemitones(reed: Reed): number {
    const sounding = plainNote(reed)
    const neighbour = plainNote({ hole: reed.hole, breath: reversed(reed.breath) })
    if (sounding >= neighbour) return 0

    return neighbour - sounding + 1
}

function plainNote(reed: Reed): MidiNote {
    return reed.breath === 'blow' ? blowNotes[reed.hole] : drawNotes[reed.hole]
}
